#!/usr/bin/env python3
"""
gRPC å®¢æˆ·ç«¯ç¤ºä¾‹
è°ƒç”¨ gRPC æœåŠ¡å™¨æä¾›çš„è®¡ç®—æœåŠ¡
"""

import grpc

# å¯¼å…¥ç”Ÿæˆçš„ gRPC ä»£ç 
import calculator_pb2
import calculator_pb2_grpc


def print_section(title):
    """æ‰“å°åˆ†éš”æ ‡é¢˜"""
    print(f"\n{'=' * 70}")
    print(f"ğŸ“Œ {title}")
    print('=' * 70)


def run_client():
    """è¿è¡Œ gRPC å®¢æˆ·ç«¯"""
    
    # åˆ›å»ºè¿æ¥é€šé“
    with grpc.insecure_channel('localhost:50051') as channel:
        # åˆ›å»ºå­˜æ ¹ï¼ˆstubï¼‰
        stub = calculator_pb2_grpc.CalculatorStub(channel)
        
        print("=" * 70)
        print("ğŸ”Œ gRPC å®¢æˆ·ç«¯")
        print("   è¿æ¥åˆ°: localhost:50051")
        print("=" * 70)
        
        # ==================== æµ‹è¯• 1: åŸºæœ¬ç®—æœ¯è¿ç®— ====================
        print_section("1. åŸºæœ¬ç®—æœ¯è¿ç®—")
        
        # åŠ æ³•
        response = stub.Add(calculator_pb2.BinaryOperation(a=15, b=27))
        print(f"[å®¢æˆ·ç«¯] Add(15, 27)")
        print(f"         ç»“æœ: {response.value}")
        print(f"         æ¶ˆæ¯: {response.message}")
        
        # å‡æ³•
        response = stub.Subtract(calculator_pb2.BinaryOperation(a=100, b=35))
        print(f"\n[å®¢æˆ·ç«¯] Subtract(100, 35)")
        print(f"         ç»“æœ: {response.value}")
        print(f"         æ¶ˆæ¯: {response.message}")
        
        # ä¹˜æ³•
        response = stub.Multiply(calculator_pb2.BinaryOperation(a=8, b=9))
        print(f"\n[å®¢æˆ·ç«¯] Multiply(8, 9)")
        print(f"         ç»“æœ: {response.value}")
        print(f"         æ¶ˆæ¯: {response.message}")
        
        # é™¤æ³•
        response = stub.Divide(calculator_pb2.BinaryOperation(a=144, b=12))
        print(f"\n[å®¢æˆ·ç«¯] Divide(144, 12)")
        print(f"         ç»“æœ: {response.value}")
        print(f"         æ¶ˆæ¯: {response.message}")
        
        # ==================== æµ‹è¯• 2: å¹³æ–¹æ ¹ ====================
        print_section("2. å¹³æ–¹æ ¹è¿ç®—")
        
        response = stub.SquareRoot(calculator_pb2.Number(value=16))
        print(f"[å®¢æˆ·ç«¯] SquareRoot(16)")
        print(f"         ç»“æœ: {response.value}")
        print(f"         æ¶ˆæ¯: {response.message}")
        
        response = stub.SquareRoot(calculator_pb2.Number(value=2))
        print(f"\n[å®¢æˆ·ç«¯] SquareRoot(2)")
        print(f"         ç»“æœ: {response.value}")
        print(f"         æ¶ˆæ¯: {response.message}")
        
        # ==================== æµ‹è¯• 3: é”™è¯¯å¤„ç† ====================
        print_section("3. é”™è¯¯å¤„ç†")
        
        # é™¤ä»¥é›¶
        print("[å®¢æˆ·ç«¯] æµ‹è¯•é™¤ä»¥é›¶:")
        try:
            response = stub.Divide(calculator_pb2.BinaryOperation(a=10, b=0))
        except grpc.RpcError as e:
            print(f"         âŒ æ•è·é”™è¯¯: {e.details()}")
            print(f"         çŠ¶æ€ç : {e.code()}")
        
        # è´Ÿæ•°å¹³æ–¹æ ¹
        print("\n[å®¢æˆ·ç«¯] æµ‹è¯•è´Ÿæ•°å¹³æ–¹æ ¹:")
        try:
            response = stub.SquareRoot(calculator_pb2.Number(value=-4))
        except grpc.RpcError as e:
            print(f"         âŒ æ•è·é”™è¯¯: {e.details()}")
            print(f"         çŠ¶æ€ç : {e.code()}")
        
        # ==================== æµ‹è¯• 4: æµå¼å“åº” ====================
        print_section("4. æµå¼å“åº” (æœåŠ¡å™¨æµ)")
        
        print("[å®¢æˆ·ç«¯] è¯·æ±‚: GetSquares(1 åˆ° 5)")
        print("         æ¥æ”¶æµå¼æ•°æ®:")
        
        # è°ƒç”¨æµå¼ RPC
        responses = stub.GetSquares(calculator_pb2.Number(value=5))
        
        for response in responses:
            print(f"         â† {response.message}")
        
        # ==================== å®Œæˆ ====================
        print_section("æµ‹è¯•å®Œæˆ")
        print("âœ… æ‰€æœ‰ gRPC è°ƒç”¨æˆåŠŸå®Œæˆï¼")
        print()
        print("ğŸ’¡ gRPC ç‰¹ç‚¹ï¼š")
        print("   - ä½¿ç”¨ Protocol Buffers å®šä¹‰æ¥å£")
        print("   - æ”¯æŒå¤šç§è°ƒç”¨æ¨¡å¼ï¼ˆä¸€å…ƒã€æµå¼ï¼‰")
        print("   - é«˜æ€§èƒ½ã€å¼ºç±»å‹")
        print("   - è·¨è¯­è¨€æ”¯æŒ")
        print("=" * 70)


if __name__ == '__main__':
    try:
        run_client()
    except grpc.RpcError as e:
        print(f"\nâŒ gRPC é”™è¯¯: {e.details()}")
        print(f"   çŠ¶æ€ç : {e.code()}")
        print("\nğŸ’¡ æç¤º: è¯·ç¡®ä¿ gRPC æœåŠ¡å™¨æ­£åœ¨è¿è¡Œ")
        print("   å¯åŠ¨å‘½ä»¤: python3 grpc_server.py")
    except Exception as e:
        print(f"\nâŒ é”™è¯¯: {e}")

